### PR TITLE
Fix issues with coherent

### PR DESCRIPTION
**Pull request by piotrAMD**
Make image load/store respect "coherent".

Set memory as "coherent" if accesses are decorated with NonPrivatePointer
or ImageOperandsNonPrivateTexel.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8c17686228d157a1fa67dff10c22ba932a54190a
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 8b2381a1d05cd2c9c4e7cc2eeda053e76d8c9a4a</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7b14353794b408b8d7fd87ae9a9f4f553e6d4782</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 955/955 (100.0%)</li>
<li>Failed: 0/955 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 40467/40468 (100.0%)</li>
<li><details><summary>Failed: 1/40468 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.instruction.graphics.image_sampler.imagesample.combined_image_sampler_separate_variables.optypeimage_mismatch.rgba8i.shader_tesse
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/ImageBuilder.cpp [hunk 1]
```diff
@@ -465,7 +465,15 @@ Value *ImageBuilder::CreateImageLoad(Type *resultTy, unsigned dim, unsigned flag
     imageDescArgIndex = args.size();
     args.push_back(imageDesc);
     args.push_back(getInt32(tfe));
-    args.push_back(getInt32(((flags & ImageFlagCoherent) ? 1 : 0) | ((flags & ImageFlagVolatile) ? 2 : 0)));
+
+    // glc/slc/dlc bits
+    unsigned cachePolicyMask = (flags & ImageFlagVolatile) ? 2 : 0; // slc
```
Comment (amdrexu) [inline lgc/builder/ImageBuilder.cpp:470]: Could we add the magic numbers to builder.h (// Image operations)
enum {
  CachePolicyNone = 0,
  CachePolicyGlc =0x1,
  CachePolicySlc = 0x2,
  CachePolicyGlc = 0x4,
}

Comment (trenouf) [reply lgc/builder/ImageBuilder.cpp:470]: Or to IntrinsDefs.h

Comment (nhaehnle) [inline lgc/builder/ImageBuilder.cpp:470]: I don't think mapping `volatile` to SLC is correct. SLC mostly affects the allocation policy of L2, basically LRU vs. streaming -- LLVM by default sets SLC for non-temporal ops. That's not the same as volatile.

I don't think there's a reason for volatile to behave any different from coherent.

Comment (piotrAMD) [reply lgc/builder/ImageBuilder.cpp:470]: Thanks, I will have a look at "volatile". Note: my patch does not change the behaviour w.r.t. volatile, I've only added "dlc" bit for "coherent".

Comment (piotrAMD) [reply lgc/builder/ImageBuilder.cpp:470]: Good point, will add them.

Comment (amdrexu) [reply lgc/builder/ImageBuilder.cpp:470]: Yes, IntrinsDefs.h has a similar definition CoherentFlag though it is not used in builder (just used in patch).

**Comment by piotrAMD** [comment]
Updated the patch with a clang formatting fix. The error in CTS looks spurious. OK to merge?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 90c4c5bacc253d363e1c03d315c59aa6a3849f22
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev b5e6dde2a0aa5dd2ce6d794e58d89ef715570f2d</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 5ae05ed8bd4f232c8580db96c021cad1e5c300e3</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 955/955 (100.0%)</li>
<li>Failed: 0/955 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/ImageBuilder.cpp [hunk 2]
```diff
@@ -626,7 +634,13 @@ Value *ImageBuilder::CreateImageStore(Value *texel, unsigned dim, unsigned flags
...
     args.push_back(imageDesc);
     args.push_back(getInt32(0)); // tfe/lwe
-    args.push_back(getInt32(((flags & ImageFlagCoherent) ? 1 : 0) | ((flags & ImageFlagVolatile) ? 2 : 0)));
+
+    // glc bit
+    CoherentFlag coherent = {};
+    if (flags & (ImageFlagCoherent | ImageFlagVolatile)) {
+      coherent.bits.glc = true;
+    }
+    args.push_back(getInt32(coherent.u32All));
```
Comment (nhaehnle) [inline lgc/builder/ImageBuilder.cpp:643]: This shouldn't be necessary.

**Pull request merged**